File: LeetCode/Medium/Replace Words/Go/replace_words.go
package main

import (
	"fmt"
	"strings"
)

type Trie struct {
	children [26]*Trie
	word     string
}

func New() *Trie { return &Trie{} }

func (this *Trie) Insert(word string) {
	root := this
	for i := 0; i < len(word); i++ {
		index := int(word[i]) - int('a')
		if root.children[index] == nil {
			root.children[index] = New()
		}
		root = root.children[index]
	}
	root.word = word
}

func (this *Trie) Search(word string) string {
	root := this
	for i := 0; i < len(word); i++ {
		index := int(word[i]) - int('a')
		if len(root.word) > 0 {
			return root.word
		}
		if root.children[index] == nil {
			return ""
		}
		root = root.children[index]
	}
	return root.word
}

func replace_words(dict []string, sen string) string {
	trie := New()
	for i := 0; i < len(dict); i++ {
		trie.Insert(dict[i])
	}
	str := strings.Split(sen, " ")
	for i := 0; i < len(str); i++ {
		curr := str[i]
		v := trie.Search(curr)
		if len(v) > 0 {
			str[i] = v
		}
	}
	return strings.Join(str, " ")
}

func main() {
	fmt.Println(replace_words([]string{"cat", "bat", "rat"}, "the cattle was rattled by the battery"))
}

File: LeetCode/Medium/Replace Words/c++/replace_words.cc
struct Node {
    string word;
    Node* children[26];
};

class Trie {
    Node* t;

public:
    Trie() {
        t = new Node();
    }

    string search(string word) {
        Node* root = this->t;
        int ind;

        for (int i = 0; i < word.length(); i++) {
            ind = word[i] - 'a';
            if (root->children[ind] == nullptr) break;
            if (root->word != "") return root->word;
            root = root->children[ind];
        }

        return root->word;

    }

    void insert(string word) {
        Node* root = this->t;
        int ind;

        for (int i = 0; i < word.length(); i++) {
            ind = word[i] - 'a';
            if (root->children[ind] == nullptr) root->children[ind] = new Node();
            root = root->children[ind];
        }
        root->word = word;
    }

};

class Solution {
public:
    string replaceWords(vector<string>& dictionary, string sentence) {
        Trie *trie = new Trie();
        stringstream ss(sentence);
        vector<string> words;
        string s, temp;

        for (string s : dictionary) trie->insert(s);
        while (getline(ss, s, ' ')) words.push_back(s);

        s = "";
        for (int i = 0; i < words.size(); i++) {
            temp = trie->search(words[i]);
            if (temp == "") s += words[i];
            else s += temp;
            s += " ";
        }

        return s.substr(0, s.length()-1);
    }
};

